### DISCUSSION TITLE
Simulate IKeyBindingHandler press and release event

### DISCUSSION THREAD

**Discussion by paring-chan** [ppy/osu-framework#4970] [Q&A]
same as title
I'm going to make touchscreen controls for osu rulesets

**Comment by bdach** [comment]
you mean something like [this](https://github.com/ppy/osu-framework/blob/4882f584d6ede9f294bf9cf33872f42f08c7d5c6/osu.Framework/Testing/Input/ManualInputManager.cs#L103-L116), but for arbitrary `IKeyBindingHandler<T>`?

keep in mind that you can always do

```csharp
((IKeyBindingHandler<T>)handler).OnPressed(action);
```

but i suppose it would be simple enough to expose a helper for this.

**Reply by paring-chan** [reply]
I have to send keys to make touchscreen controls for osu

**Reply by bdach** [reply]
i'm not sure i follow. what are you trying to achieve? are you implementing https://github.com/ppy/osu/issues/4201?

if that is the case then i'm not necessarily sure "keys" need to be or should be "sent". consider that you could potentially unbind all keyboard bindings and the virtual controls should continue to work.

**Reply by bdach** [reply]
also `IKeyBindingContainer.TriggerPressed()` exists:

https://github.com/ppy/osu-framework/blob/49c954321c3686628b2c223670363438f88a0341/osu.Framework/Input/Bindings/KeyBindingContainer.cs#L360-L368

the press/release must be fired on the container rather than the handlers, because the handlers receive the event from the container. the container knows the correct way to route the action to the proper handler.

**Reply by paring-chan** [reply]
I'll try it. Thanks.
